File: internal/model/contract_test.go
// Contract test: the checked-in golden snapshot must parse into this package's types,
// pinning the wire shape shared with the frontend.

package model

import (
	"encoding/json"
	"os"
	"testing"
)

// TestGoldenSnapshotContract guards the Go<->TS wire contract: both sides parse
// the SAME checked-in fixture (testdata/contract/golden_snapshot.json at the repo
// root) and must agree on every field. A silent field-name/shape drift between
// this package and web/src/state/race.ts's mirrored types fails here, in CI,
// rather than as a runtime "undefined" deep in the frontend.
func TestGoldenSnapshotContract(t *testing.T) {
	raw, err := os.ReadFile("../../testdata/contract/golden_snapshot.json")
	if err != nil {
		t.Fatal(err)
	}
	var s Snapshot
	if err := json.Unmarshal(raw, &s); err != nil {
		t.Fatalf("golden fixture does not decode as model.Snapshot: %v", err)
	}

	if s.SessionKey != "contract-test" || s.Mode != ModeReplay || s.Label != "Contract Fixture" {
		t.Fatalf("top-level identity fields mismatched: %+v", s)
	}
	if s.Rev != 42 || s.TimeMs != 3300000 {
		t.Fatalf("rev/timeMs mismatched: rev=%d timeMs=%d", s.Rev, s.TimeMs)
	}
	if s.TotalLaps != 53 {
		t.Fatalf("totalLaps mismatched: %d", s.TotalLaps)
	}
	if len(s.Track) != 2 || s.Track[1].X != 0.9 {
		t.Fatalf("track mismatched: %+v", s.Track)
	}

	if len(s.Cars) != 2 {
		t.Fatalf("want 2 cars, got %d: %+v", len(s.Cars), s.Cars)
	}
	c1 := s.Cars[1]
	if c1.DriverNum != 1 || c1.Code != "VER" || c1.Team != "Red Bull" || c1.Status != StatusOnTrack {
		t.Fatalf("car 1 identity/status mismatched: %+v", c1)
	}
	if c1.Lap != 12 {
		t.Fatalf("car 1 lap mismatched: %+v", c1)
	}
	if c1.Tyre != "SOFT" || c1.TyreAge != 5 || c1.LastLapMs != 81234 || c1.BestLapMs != 80950 {
		t.Fatalf("car 1 timing fields mismatched: %+v", c1)
	}
	if c1.Speed != 312 || c1.Gear != 7 || c1.Throttle != 100 || c1.DRS != true {
		t.Fatalf("car 1 telemetry fields mismatched: %+v", c1)
	}
	if s.Cars[44].Status != StatusPit {
		t.Fatalf("car 44 status mismatched: %+v", s.Cars[44])
	}

	if len(s.Messages) != 1 || s.Messages[0].Message != "GREEN FLAG" || s.Messages[0].Category != "Flag" {
		t.Fatalf("messages mismatched: %+v", s.Messages)
	}
	if len(s.Radio) != 1 || s.Radio[0].DriverNum != 1 || s.Radio[0].Clip == "" {
		t.Fatalf("radio mismatched: %+v", s.Radio)
	}
	if len(s.LapTrace[1]) != 4 {
		t.Fatalf("lapTrace mismatched: %+v", s.LapTrace)
	}

	if len(s.Stints[1]) != 2 || s.Stints[1][1].Compound != "HARD" || s.Stints[1][1].StartLap != 15 {
		t.Fatalf("stints mismatched: %+v", s.Stints)
	}
	if len(s.PitStops[1]) != 1 || s.PitStops[1][0].Lap != 14 || s.PitStops[1][0].DurationS != 23.4 {
		t.Fatalf("pitStops mismatched: %+v", s.PitStops)
	}
	if s.Weather == nil || s.Weather.TrackTempC != 41.2 || s.Weather.AirTempC != 28.5 || s.Weather.Rainfall {
		t.Fatalf("weather mismatched: %+v", s.Weather)
	}
}

File: web/src/state/contract.test.ts
// Contract test: the golden snapshot fixture must apply cleanly through the frontend
// reducer.

import { readFileSync } from 'node:fs';
import { describe, it, expect } from 'vitest';
import { emptyState, applyMessage } from './race';

// Guards the Go<->TS wire contract: both sides parse the SAME checked-in fixture
// (testdata/contract/golden_snapshot.json at the repo root) and must agree on
// every field. See internal/model/contract_test.go for the Go half. A silent
// field-name/shape drift between model.go and this file's mirrored types fails
// here, in CI, rather than as a runtime "undefined" deep in the frontend.
describe('golden snapshot contract', () => {
  it('parses the same checked-in fixture the Go side asserts against', () => {
    const raw = readFileSync('../testdata/contract/golden_snapshot.json', 'utf-8');
    const data = JSON.parse(raw);
    const s = applyMessage(emptyState(), { type: 'snapshot', data });

    expect(s.session).toBe('contract-test');
    expect(s.mode).toBe('replay');
    expect(s.label).toBe('Contract Fixture');
    expect(s.rev).toBe(42);
    expect(s.timeMs).toBe(3300000);
    expect(s.totalLaps).toBe(53);
    expect(s.track).toHaveLength(2);
    expect(s.track[1]).toEqual({ x: 0.9, y: 0.8 });
    expect(s.corners).toHaveLength(1);
    expect(s.corners[0]).toEqual({ number: 1, x: 0.1, y: 0.2 });

    expect(Object.keys(s.cars)).toHaveLength(2);
    expect(s.cars[1].driverNum).toBe(1);
    expect(s.cars[1].code).toBe('VER');
    expect(s.cars[1].lap).toBe(12);
    expect(s.cars[1].status).toBe('OnTrack');
    expect(s.cars[1].tyre).toBe('SOFT');
    expect(s.cars[1].tyreAge).toBe(5);
    expect(s.cars[1].drs).toBe(true);
    expect(s.cars[44].status).toBe('Pit');

    expect(s.messages).toHaveLength(1);
    expect(s.messages[0].message).toBe('GREEN FLAG');
    expect(s.radio).toHaveLength(1);
    expect(s.radio[0].driverNum).toBe(1);
    expect(s.lapTrace[1]).toHaveLength(4);

    expect(s.stints[1]).toHaveLength(2);
    expect(s.stints[1][1].compound).toBe('HARD');
    expect(s.stints[1][1].startLap).toBe(15);
    expect(s.pitStops[1]).toHaveLength(1);
    expect(s.pitStops[1][0].lap).toBe(14);
    expect(s.pitStops[1][0].durationS).toBe(23.4);
    expect(s.pedalTraces[1].throttle).toEqual([0, 40, 100, 100]);
    expect(s.pedalTraces[1].brake).toEqual([0, 60, 0, 0]);
    expect(s.pedalTraces[1].gear).toEqual([1, 3, 7, 7]);
    expect(s.sectorDominance).toEqual([1]);
    expect(s.weather?.trackTempC).toBe(41.2);
    expect(s.weather?.airTempC).toBe(28.5);
    expect(s.weather?.rainfall).toBe(false);
  });
});
